### PR TITLE
[ENH] Add status command to display high-level participant counts

### LINKED ISSUE #307: [ENH] Add `nipoppy info` (or `nipoppy status`) command
Which will print the following information to the console in a pretty way (with nice colours):
- Dataset name and available visit/session IDs
- Number of participants in manifest
  - Possibly also number of participants with imaging data in manifest (+ maybe breakdown of BIDS datatypes)
- Doughnut information if available
- Bagel information if available

### PR DESCRIPTION
**Pull request by nikhil153**
- Closes #307

Changes proposed in this pull request:
- add nipoppy command: `status` to show aggregate participant counts in `manifest`, `doughnut`, and `imaging_bagel`

## Checklist (for reviewers)
_This section is for the PR reviewer_

- [x] PR has an interpretable title with a prefix (e.g. `[BUG]`, `[DOC]`, `[ENH]`, `[MAINT]`)\
Refer to [NumPy Development Guide](https://numpy.org/doc/stable/dev/development_workflow.html#writing-the-commit-message) for a full list
- [x] PR links to GitHub issue with mention `Closes #XXXX`
- [x] Tests pass
- [ ] Checks pass

For new features:
- [ ] Tests have been added

For bug fixes:
- ~There is at least one test that would fail under the original bug conditions~

### CODE REVIEW

File: nipoppy/workflows/dataset_status.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+        self.status_df = pd.concat([self.status_df, bagel_status_df], axis=1)
+
+    def df_to_table(self):
+        """Convert a pandas.DataFrame obj into a rich.Table obj."""
+        console = Console()
+
+        # Initiate a Table instance
+        title = "Participant counts at each nipoppy checkpoint"
+        table = Table(show_header=True, header_style="bold red", title=title)
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:202]: I'm not a huge fan of the big space between the title and the header row, is it possible to reduce it?

File: nipoppy/workflows/dataset_status.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        doughnut_status_df = doughnut.groupby([doughnut.col_session_id]).sum()[
+            [
+                doughnut.col_in_pre_reorg,
+                doughnut.col_in_post_reorg,
+                doughnut.col_in_bids,
+            ]
+        ]
+
+        self.logger.debug(f"doughnut_status_df: {doughnut_status_df}")
+        self.status_df = pd.concat([self.status_df, doughnut_status_df], axis=1)
```
[the remark this answers is not in the document]
Comment (michellewang) [reply nipoppy/workflows/dataset_status.py:139]: Same thing for the other instances of this too

File: nipoppy/workflows/dataset_status.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        super().__init__(
+            dpath_root=dpath_root,
+            name="init",
+            fpath_layout=fpath_layout,
+            logger=logger,
+            dry_run=dry_run,
+        )
+        self.fname_readme = "README.md"
+        self.save_status_to_disk = save_status_to_disk  # TODO maybe
+        self.status_df = pd.DataFrame()
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:46]: Edge case but I am wondering if it would be better to define the empty `status_df` here or in a `run_setup` method. Not sure...

**Review by michellewang — CHANGES_REQUESTED** [review]
Looks good @nikhil153, and worked when I tried it locally!

I think the main thing would be to add a test and to fix the `pre-commit` [errors](https://results.pre-commit.ci/run/github/466274304/1730472071.tM81f1yXRoemCxamKDPzHA). Also decide what to do with the saving-to-disk option

File: nipoppy/cli/parser.py [hunk 4]
```diff
@@ -188,6 +189,31 @@ def add_subparser_init(
...
+        COMMAND_STATUS,
+        description=description,
+        help=description,
+        formatter_class=formatter_class,
+        add_help=False,
+    )
+
+    parser = add_arg_dataset_root(parser)
+
+    parser.add_argument(
```
[the remark this answers is not in the document]
Comment (nikhil153) [reply nipoppy/cli/parser.py:208]: I thought of that but then it makes it harder to report `diff` counts..

File: nipoppy/workflows/dataset_status.py (hunk 3 above)
Comment (nikhil153) [reply nipoppy/workflows/dataset_status.py:39]: fixed!

File: nipoppy/workflows/dataset_status.py (hunk 1 above)
Comment (nikhil153) [reply nipoppy/workflows/dataset_status.py]: tired and failed..

**Comment by nikhil153** [comment]
@michellewang - I think i have addressed all the comments, added the test and most importantly emojis!

File: nipoppy/workflows/dataset_status.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+class StatusWorkflow(BaseWorkflow):
+    """Workflow for status command."""
+
+    def __init__(
+        self,
+        dpath_root: Path,
+        fpath_layout: Optional[StrOrPathLike] = None,
+        logger: Optional[logging.Logger] = None,
+        dry_run: bool = False,
+        save_status_to_disk: bool = False,
```
[the remark this answers is not in the document]
Comment (nikhil153) [reply nipoppy/workflows/dataset_status.py]: Commented out for now.

File: nipoppy_cli/docs/source/personas.md
Comment (michellewang) [inline nipoppy_cli/docs/source/personas.md:1]: This file does not belong in this PR!

File: nipoppy/cli/run.py
Comment (michellewang) [inline nipoppy/cli/run.py:1]: To make Codecov happy we will need to add tests for the changes in the CLI too (`run.py` and `parser.py`)

File: nipoppy/workflows/dataset_status.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+
+        # Get the number of imaging sessions in the manifest
+        session_ids = imaging_manifest[manifest.col_session_id].unique()
+
+        self.logger.info(
+            f"\tNumber of participants (imaging and non-imaging): "
+            f"{len(participant_ids)}"
+        )
+        self.logger.info(
+            f"\tAvailable  visits (imaging and non-imaging) (n={len(visit_ids)}): "
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:100]:
```suggestion
            f"\tAvailable visits (imaging and non-imaging) (n={len(visit_ids)}): "
```

(extra space)

File: nipoppy/workflows/dataset_status.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+        sparkles_emoji = "✨"
+        rocket_emoji = "🚀"
+        bagel_emoji = "🥯"
+        party_popper_emoji = "🎉"
+        confetti_emoji = "🎊"
+
+        # based on quantity
+        high_throughput = 100
+        bento_box_emoji = "🍱"  # 100 subjects in BIDS
+        mate_emoji = "🧉"  # 100 subjects processed
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:232]: I think it is confusing to have secret emojis 😅

File: nipoppy/workflows/dataset_status.py [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+        status_df = pd.concat([status_df, bagel_status_df], axis=1)
+        return status_df
+
+    def _add_rewards(self, status_df: pd.DataFrame) -> pd.DataFrame:
+        """Add rewards column to the status DataFrame."""
+        status_df = status_df.fillna(0).astype(int).reset_index()
+        status_df = status_df.sort_values(by=self.manifest.col_session_id)
+
+        # generate positive reinforcement emojis
+        clipboard = "📋"
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:218]: Should this be added to the legend?

File: nipoppy/workflows/dataset_status.py (hunk 7 above)
Comment (michellewang) [reply nipoppy/workflows/dataset_status.py:232]: Also right now we have e.g. the broom emoji and the bento box emoji simultaneously -- are they supposed to be either/or or cumulative?

File: .flake8 [hunk 9]
```diff
@@ -14,7 +14,7 @@ per-file-ignores =
 	tests/test_workflow_*.py: D100, D103, E127, E131, E501
     docs/source/conf.py: E501
 docstring-convention = numpy
-max_complexity = 12
+max_complexity = 15
```
[the remark this answers is not in the document]
Comment (nikhil153) [reply .flake8:17]: yup for the rewards criteria function. It feels silly to create separate function of each of them.

File: nipoppy/workflows/dataset_status.py [hunk 10]
```diff
@@ -88,44 +93,52 @@ def check_manifest(self):
...
+        self.logger.info(f"Status at nipoppy_checkpoint: {nipoppy_checkpoint}")
 
-        if not self.layout.fpath_doughnut.exists():
-            self.logger.info(f"No doughnut file found at {self.layout.fpath_doughnut}.")
+        try:
+            doughnut = self.doughnut
+        except FileNotFoundError:
+            self.logger.warning(
+                f"No doughnut file found at {self.layout.fpath_doughnut}."
+            )
```
[the remark this answers is not in the document]
Comment (nikhil153) [reply nipoppy/workflows/dataset_status.py:131]: Does that happen any time self.doughnut is called? I guess i was thinking of a scenario when someone accidentally deletes the doughnut.

File: nipoppy/workflows/dataset_status.py (hunk 7 above)
Comment (nikhil153) [reply nipoppy/workflows/dataset_status.py:232]: You don't like secret restaurant menus? 
they were supposed to be cumulative, but happy to think to avoid too many emojis.

File: nipoppy/workflows/dataset_status.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (michellewang) [reply nipoppy/workflows/dataset_status.py:131]: It happens the first time `self.doughnut` is called, where it either loads the doughnut or regenerates a new one. If someone deletes the doughnut while `StatusWorkflow` is running then there is nothing we can do against that, but if they delete it before running `nipoppy status` then the doughnut will be regenerated

File: nipoppy/workflows/dataset_status.py (hunk 7 above)
Comment (michellewang) [reply nipoppy/workflows/dataset_status.py:232]: What if we have the emojis next to the numbers inside the cells?

For example something like this:
```
| session_id | in_manifest | in_pre_reorg | in_post_reorg | in_bids | fmriprep | mriqc |
|------------|-------------|--------------|---------------|---------|----------|-------|
| 1          | 100 📋      | 100 🍩       | 100 🍩        | 100 ✨  | 100 🎉   | 100 🎉|
| 2          | 90 📋       | 80 🍩        | 80 🍩         | 70 🧹   | 60 🏃    | 70 🏃 |
```

We would be losing the bagel emoji and the 🎊 emoji though. Also a legend might not be necessary anymore? But not sure

File: nipoppy/cli/parser.py [hunk 11]
```diff
@@ -188,6 +189,33 @@ def add_subparser_init(
...
+
+    parser = add_arg_dataset_root(parser)
+
+    # TODO in future release
+    # parser.add_argument(
+    #     "--save-status-to-disk",
+    #     action="store_true",
+    #     help=("Save the status to disk."
+    #           " This will overwrite the existing status file."),
+    # )
```
Comment (michellewang) [inline nipoppy/cli/parser.py:214]: I am fine with commenting these out but my top preference would just be to delete these lines and open another issue for a future PR...

File: nipoppy/workflows/dataset_status.py [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+        bagel_status_df = bagel_pipeline_df.groupby([bagel.col_session_id]).count()
+
+        self.logger.debug(f"bagel_status_df: {bagel_status_df}")
+
+        status_df = pd.concat([status_df, bagel_status_df], axis=1)
+
+        bagel_cols = list(status_df.columns[status_df.columns.str.contains("\n")])
+        return status_df, bagel_cols
+
+    def _add_manifest_rewards(
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:273]: It might be good to have some comments explaining the `if`/`elif`/`else` conditions. For all three reward conditions

**Review by michellewang — COMMENTED** [review]
Almost there!

File: nipoppy/workflows/dataset_status.py [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+legend_dict = {
+    clipboard_emoji: "Manifest entries available!",
+    sad_cat_emoji: "No available data are curated or processed",
+    broom_emoji: "Data clean-up (i.e. BIDSification) has started!",
+    doughnut_emoji: "Curation in progress... Doughnut is made!",
+    sparkles_emoji: "BIDSification is complete! Tidiness sparks joy!",
+    rocket_emoji: "Data processing has launched successfully!",
+    bagel_emoji: "Data processing in progress... Bagels are baking!",
+    party_popper_emoji: "At least 1 pipeline processing is complete!",
+    confetti_emoji: "All data are curated and processed. Time to celebrate!",
```
[the remark this answers is not in the document]
Comment (nikhil153) [reply nipoppy/workflows/dataset_status.py:50]: Actually, I am thinking of a three-step emoji process: start (broom / rocket) --> progress (doughnut / bagel) --> success (sparkles / party_popper). Both broom and rocket have the least stringent criteria.

File: nipoppy/cli/parser.py (hunk 11 above)
Comment (nikhil153) [reply nipoppy/cli/parser.py:214]: okay i will remove it

File: tests/test_workflow_dataset_status.py [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+from .conftest import get_config
+
+
+@pytest.fixture(params=["my_dataset", "dataset_dir"])
+def dpath_root(request: pytest.FixtureRequest, tmp_path: Path) -> Path:
+    return tmp_path / request.param
+
+
+def test_run(dpath_root: Path):
```
[the remark this answers is not in the document]
Comment (nikhil153) [reply tests/test_workflow_dataset_status.py:20]: The high-throughput ones will be hard to generate... but I will try.

File: nipoppy/workflows/dataset_status.py [hunk 15]
```diff
@@ -0,0 +1,466 @@
...
+broom_emoji = "🧹"
+doughnut_emoji = "🍩"
+sparkles_emoji = "✨"
+rocket_emoji = "🚀"
+bagel_emoji = "🥯"
+party_popper_emoji = "🎉"
+confetti_emoji = "🎊"
+
+# based on quantity
+medium_throughput = 100
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:32]: @mathdugre wants `throughput` to be renamed `volume`

File: tests/test_workflow_dataset_status.py [hunk 16]
```diff
@@ -0,0 +1,571 @@
...
+
+
+def make_status_df(
+    n_participants,
+    session_ids=("BL", "M12", "M24"),
+    pipeline_configs=(
+        ("dcm2bids", "1.0.0", "prepare"),
+        ("dcm2bids", "1.0.0", "convert"),
+        ("fmriprep", "1.0.0", "default"),
+    ),
```
Comment (michellewang) [inline tests/test_workflow_dataset_status.py:241]: I think this function is implementing too much of the logic that is being used in the tested code. My preference would be to generate the `status_df` as directly as possible, e.g.

```python
def make_status_df(
    session_ids=("BL", "M12", "M24"),
    counts_by_col=(
        ("in_manifest", (20, 20, 19)),
        ("in_pre_reorg", (18, 20, 19)),
        ("in_post_reorg", (18, 18, 18)),
        ("in_bids", (18, 18, 18)),
        ("fmriprep\n1.0.0\ndefault", (0, 5, 5)),
        ("freesurfer\n1.0.0\ndefault", (18, 18, 18)),
    )
):
    df = pd.DataFrame(session_ids, columns=[Manifest.col_session_id])
    for col, counts in counts_by_col:
        df[col] = counts
    return df
```

Then for the test case, we could have parametrization like this:

```python
@pytest.mark.parametrize(
    ",session_ids,counts_by_col,emojis",
    [
        (
            ("BL", "M12"),
            (
                ("in_manifest", (20, 20)),
                ("in_pre_reorg", (18, 0)),
                ("in_post_reorg", (18, 0)),
                ("in_bids", (18, 0)),
                ("fmriprep\n1.0.0\ndefault", (20, 0)),
                ("freesurfer\n1.0.0\ndefault", (18, 0)),
            ),
            (
                {
                    "init\nrewards": [clipboard_emoji],
                    "curation\nrewards": [doughnut_emoji, broom_emoji],
                    "processing\nrewards": [bagel_emoji, party_popper_emoji],
                },
                {
                    "init\nrewards": [sad_cat_emoji],
                    "curation\nrewards": [],
                    "processing\nrewards": [],
                },
            ),
        ),
    ],
)

```

File: nipoppy/workflows/dataset_status.py [hunk 17]
```diff
@@ -0,0 +1,467 @@
...
+slow_sloth_emoji = "🦥"
+quick_tiger_emoji = "🐅"
+
+secret_rewards_dict = {
+    bento_box_emoji: "1000+ subjects in BIDS! That's impressive attention to detail!",
+    mate_emoji: "1000+ subjects are processed. Deserves a drink!",
+    night_owl_emoji: "Night owl! That's some dedication!",
+    slow_sloth_emoji: "Slow sloth! This dataset could use some attention!",
+    quick_tiger_emoji: "Quick tiger! You are blazing through!",
+    surprise_box_emoji: "There are more secret rewards to be unlocked!",
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:51]: Delete/comment out? 😅

File: nipoppy/workflows/dataset_status.py [hunk 18]
```diff
@@ -0,0 +1,467 @@
...
+from rich.console import Console
+from rich.table import Table
+
+from nipoppy.env import LogColor, StrOrPathLike
+from nipoppy.tabular.bagel import STATUS_SUCCESS
+from nipoppy.workflows.base import BaseWorkflow
+
+# generate positive reinforcement emojis for data curation and processing
+# three-step process (least stringent to most stringent)
+# 1. start: (broom / rocket)
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:18]: How about changing "rocket" to "hourglass"? ⏳ I feel like that corresponds better with broom

File: nipoppy/workflows/dataset_status.py [hunk 19]
```diff
@@ -0,0 +1,467 @@
...
+    sad_cat_emoji: "No available data are curated or processed",
+    broom_emoji: "Data clean-up (i.e. BIDSification) has started!",
+    doughnut_emoji: "Curation in progress... Doughnut is made!",
+    sparkles_emoji: "BIDSification is complete! Tidiness sparks joy!",
+    rocket_emoji: "Data processing has launched successfully!",
+    bagel_emoji: "Data processing in progress... Bagels are baking!",
+    party_popper_emoji: "At least 1 pipeline processing is complete!",
+    confetti_emoji: "Amazing! All data are curated and processed. Time to celebrate!",
+    sushi_emoji: "100+ subjects in BIDS! Meticulousness should be rewarded!",
+    shaved_ice_emoji: "100+ subjects are processed. Deseves a cool treat!",
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:66]: Delete/comment out?

File: nipoppy/workflows/dataset_status.py [hunk 20]
```diff
@@ -0,0 +1,467 @@
...
+        super().__init__(
+            dpath_root=dpath_root,
+            name="status",
+            fpath_layout=fpath_layout,
+            logger=logger,
+            dry_run=dry_run,
+        )
+        self.col_pipeline = "pipeline"
+
+    def run_main(self):
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:90]: We should have a test for `StatusWorkflow.run()`. And then we can check that some things like dataset name are being logged

File: nipoppy/workflows/dataset_status.py [hunk 21]
```diff
@@ -0,0 +1,467 @@
...
+        all_rewards = status_df[reward_cols].values.flatten()
+        # split the multiple emojis into a list
+        all_rewards = [reward.split() for reward in all_rewards]
+        all_rewards = [reward for sublist in all_rewards for reward in sublist] + [
+            surprise_box_emoji
+        ]
+        all_rewards = set(all_rewards)
+
+        for reward in all_rewards:
+            if reward in secret_rewards_dict.keys():
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:159]: I thought we were not doing the secret rewards anymore!

File: nipoppy/workflows/dataset_status.py [hunk 22]
```diff
@@ -0,0 +1,467 @@
...
+
+        for row_obj in status_df.iterrows():
+            row = row_obj[1]
+            reward_emojis = []
+            # doughnut emoji if at least one participant has been bidsified
+            # can't rely on other columns if we begin with BIDS dataset
+            if (row[bids_col] > 0) & (row[bids_col] <= row[manifest_col]):
+                reward_emojis.append(doughnut_emoji)
+            # sparkles emoji if all participants have been bidsified
+            if row[bids_col] == row[manifest_col]:
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:342]: Here if `bids_col` and `in_manifest` are both 0 they would get a sparkle too? Maybe we should make it so that if the manifest has nothing then no emojis are given? And then we should have a test case for that!

File: nipoppy/workflows/dataset_status.py [hunk 23]
```diff
@@ -0,0 +1,467 @@
...
+            #     reward_emojis.append(sushi_emoji)
+
+            status_df.loc[row_obj[0], reward_col_name] = " ".join(reward_emojis)
+
+        return status_df
+
+    def _add_bagel_rewards(
+        self, status_df: pd.DataFrame, status_col_dict: dict, reward_col_name: str
+    ) -> pd.DataFrame:
+        """Add begal rewards column to the status DataFrame."""
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:361]:
```suggestion
        """Add bagel rewards column to the status DataFrame."""
```

I'm surprised Codespell didn't catch that...

File: nipoppy/workflows/dataset_status.py [hunk 24]
```diff
@@ -0,0 +1,467 @@
...
+        manifest_col = status_col_dict["manifest"]
+        doughnut_cols = status_col_dict["doughnut"]
+        bagel_cols = status_col_dict["bagel"]
+
+        status_count_cols = [manifest_col] + doughnut_cols + bagel_cols
+        status_df[status_count_cols] = (
+            status_df[status_count_cols].fillna(0).astype(int)
+        )
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:369]: Can we put the `fillna(0)` calls right after `pd.concat` in the `_check_{manifest,doughnut,bagel}` methods?

File: nipoppy/workflows/dataset_status.py [hunk 25]
```diff
@@ -0,0 +1,467 @@
...
+
+    def _add_bagel_rewards(
+        self, status_df: pd.DataFrame, status_col_dict: dict, reward_col_name: str
+    ) -> pd.DataFrame:
+        """Add begal rewards column to the status DataFrame."""
+        manifest_col = status_col_dict["manifest"]
+        doughnut_cols = status_col_dict["doughnut"]
+        bagel_cols = status_col_dict["bagel"]
+
+        status_count_cols = [manifest_col] + doughnut_cols + bagel_cols
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:366]: Does it really make sense to have `doughnut_cols`? I get why we would want `in_bids` but since they might not be starting with a BIDS dataset then they won't have the other columns

File: nipoppy/workflows/dataset_status.py [hunk 26]
```diff
@@ -0,0 +1,467 @@
...
+        return status_df
+
+    def _df_to_table(self, status_df: pd.DataFrame, legend_dict: dict):
+        """Convert a pandas.DataFrame obj into a rich.Table obj."""
+        df = status_df.copy().reset_index()
+        df = df.sort_values(by=self.manifest.col_session_id)
+
+        console = Console()
+
+        crossed_fingers_emoji = "🤞"
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:415]: I have had people ask what the crossed fingers meant, so this might be confusing for some people

File: nipoppy/workflows/dataset_status.py [hunk 27]
```diff
@@ -0,0 +1,467 @@
...
+        n_colors = len(column_colors)
+        n_columns = len(df.columns)
+        if n_colors < n_columns:  # cycle through the colors
+            column_colors = (
+                column_colors * (n_columns // n_colors + 1)
+                + column_colors[: n_columns % n_colors]
+            )
+
+        for column, col_color in zip(df.columns, column_colors):
+            table.add_column(
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:444]: I *think* this should work, and it would simplify the code

```suggestion
        n_colors = len(column_colors)
        for i_col, column in enumerate(df.columns):
            col_color = column_colors[i_col % n_colors]
            table.add_column(
```

File: tests/test_workflow_dataset_status.py [hunk 28]
```diff
@@ -0,0 +1,571 @@
...
+shaved_ice_emoji = "🍧"  # 100 subjects processed
+
+# other emojis
+surprise_box_emoji = "🎁"
+high_throughput = 1000
+bento_box_emoji = "🍱"  # 1000 subjects in BIDS
+mate_emoji = "🧉"  # 1000 subjects processed
+night_owl_emoji = "🦉"
+slow_sloth_emoji = "🦥"
+quick_tiger_emoji = "🐅"
```
Comment (michellewang) [inline tests/test_workflow_dataset_status.py:44]: Delete/comment out?

File: nipoppy/workflows/dataset_status.py [hunk 29]
```diff
@@ -0,0 +1,467 @@
...
+
+        1) Number of participants in manifest per BIDS datatype,
+        2) Doughnut information if available,
+        3) Bagel information if available
+        """
+        self.logger.info("Checking the status of the dataset.")
+
+        # load global_config to get the dataset name
+        dataset_name = self.config.DATASET_NAME
+        expected_sessions = self.config.SESSION_IDS
```
[the remark this answers is not in the document]
Comment (michellewang) [reply nipoppy/workflows/dataset_status.py:101]: Or: we should have the visit/session ids printed out during the manifest check follow the same order as what is in the manifest. I kind of like that better because it lets users define their own preferred order. But this would require #427 to be addressed first

**Comment by michellewang** [comment]
Summary of emoji discussion:
1. Remove crossed fingers emoji
2. Remove all `rewards` columns
3. Remove legend
4. Add double vertical lines to separate manifest/curation/processing columns
5. Add an "Achievements" section after the main table, where achievements are like:
   - "At least one session fully BIDSified! 🧹" or "All sessions have been BIDSified! ✨"
   - etc.

**Comment by michellewang** [comment]
We can split the achievements things into another PR to make reviewing easier :)

File: tests/test_workflow_dataset_status.py (hunk 16 above)
Comment (nikhil153) [reply tests/test_workflow_dataset_status.py:241]: I tried this at first, but it gets too annoying to manually adjust numbers in `status_df` when criteria for rewards is changed. I agree there is added complexity / logic in creating this test case, but I see the generative status_df and reductive reward operations as inverses of each other. To me this is different than duplicating the logic which would be problematic from testing perspective.

Comment (michellewang) [reply tests/test_workflow_dataset_status.py:231]: I guess the main problem is that it is hard to be sure that the dataframe-generating helper function is itself correct. We don’t want to write tests for our tests…

File: nipoppy/workflows/dataset_status.py [hunk 30]
```diff
@@ -183,22 +96,20 @@ def _check_manifest(self, status_df: pd.DataFrame) -> pd.DataFrame:
...
-        session_ids = imaging_manifest[manifest.col_session_id].unique()
+        session_ids = sorted(imaging_manifest[manifest.col_session_id].unique())
 
         self.logger.info(
             f"\tNumber of participants (imaging and non-imaging): "
             f"{len(participant_ids)}"
         )
         self.logger.info(
-            f"\tVisits (imaging and non-imaging) (n={len(visit_ids)}): " f"{visit_ids}"
+            f"\tVisits (imaging and non-imaging) (n={len(visit_ids)}):{visit_ids}"
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:106]:
```suggestion
            f"\tVisits (imaging and non-imaging) (n={len(visit_ids)}): {visit_ids}"
```

File: nipoppy/workflows/dataset_status.py [hunk 31]
```diff
@@ -183,22 +96,20 @@ def _check_manifest(self, status_df: pd.DataFrame) -> pd.DataFrame:
...
         )
         self.logger.info(
             f"\tNumber of participants with imaging data: "
             f"{len(imaging_participant_ids)}"
         )
-        self.logger.info(
-            f"\tImaging sessions (n={len(session_ids)}): " f"{session_ids}"
-        )
+        self.logger.info(f"\tImaging sessions (n={len(session_ids)}):{session_ids}")
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:112]:
```suggestion
        self.logger.info(f"\tImaging sessions (n={len(session_ids)}): {session_ids}")
```

File: nipoppy/workflows/dataset_status.py [hunk 32]
```diff
@@ -245,6 +161,10 @@ def _check_bagel(self, status_df: pd.DataFrame) -> pd.DataFrame:
         self.logger.debug(f"Status at nipoppy_checkpoint: {nipoppy_checkpoint}")
         bagel = self.bagel
 
+        if bagel.empty:
+            self.logger.warning("No imaging bagel file found.")
```
Comment (michellewang) [inline nipoppy/workflows/dataset_status.py:165]:
```suggestion
            self.logger.warning('No imaging bagel file found. Run "nipoppy track" to generate an imaging bagel file')
```

File: tests/test_workflow_dataset_status.py [hunk 33]
```diff
@@ -464,108 +346,14 @@ def test_bagel(
...
-                    "init\nrewards": [sad_cat_emoji],
-                    "curation\nrewards": [],
-                    "processing\nrewards": [],
-                },
-            },
-        ),
-    ],
-)
-def test_rewards(
+def test_run(
```
Comment (michellewang) [inline tests/test_workflow_dataset_status.py:349]:
```suggestion
@pytest.mark.parametrize("bagel", [pd.DataFrame(), make_bagel()[0]])
def test_run(
```

File: tests/test_workflow_dataset_status.py [hunk 34]
```diff
@@ -464,108 +346,14 @@ def test_bagel(
...
     dpath_root: Path,
```
Comment (michellewang) [inline tests/test_workflow_dataset_status.py:350]:
```suggestion
    dpath_root: Path,
    bagel: pd.DataFrame,
```

File: tests/test_workflow_dataset_status.py [hunk 35]
```diff
@@ -464,108 +346,14 @@ def test_bagel(
...
-    session_ids: list,
-    pipeline_configs: tuple,
-    emojis: dict,
 ):
-
     workflow = StatusWorkflow(dpath_root=dpath_root)
+    workflow.config = get_config()
+    workflow.manifest = make_manifest(n_participants=10)[0]
+    workflow.doughnut = pd.DataFrame()  # Checks for empty doughnut
+    workflow.bagel = pd.DataFrame()  # Checks for empty bagel
```
Comment (michellewang) [inline tests/test_workflow_dataset_status.py:356]:
```suggestion
    workflow.bagel = bagel
```

**Review by michellewang — COMMENTED** [review]
Looks good! Added some suggestions to try to make Codecov happy, if that works then I will approve the PR. And the `BaseWorkflow.bagel` coverage is addressed indirectly in the extractor PR #399 but we can write a simple test in `test_workflow_base.py` (haven't run this so not 100% sure it works):

```python
def test_bagel_empty_if_not_found(workflow: BaseWorkflow):
    assert len(workflow.bagel) == 0
```

File: tests/test_workflow_dataset_status.py (hunk 33 above)
Comment (michellewang) [reply tests/test_workflow_dataset_status.py:349]: This is and the next 2 suggestions are for increasing test coverage

**Comment by nikhil153** [comment]
Incorporated your comments and suggestions. I am not sure though that the <100% code coverage stat was coming from the previous bagel test. I believe it is from the table column coloring branching - which is a bit annoying and probably unnecessary to test. 

image

**Comment by michellewang** [comment]
@nikhil153 the tests are failing for an unrelated reason (PyBIDS just had a release and the new version is broken). This is "fixed" in #449 -- could you check/approve/merge it? Then this PR should be good to go as long as the tests pass :)
